app/scripts/lib/channel-server-client.js

```diff
@@ -0,0 +1,294 @@
+   *   "sender": {
+   *      "city": <sender's city name>,
+   *      "country": <sender's country name>,
+   *      "region": <sender's region>,
+   *      "ua": <sender's user agent string>
+   *   }
+   * }
+   * ```
+   *
+   * The encrypted message is expected to have the following format:
+   *
+   * ```
+   * {
+   *   "message": <message name>,
+   *   "data": <any, specific to the message>
+   * }
+   *
+   * @param {Object} event
+   * @private
+   */
+  _encryptedMessageHandler (event) {
+    const { message: ciphertext, sender } = JSON.parse(event.data);
+    this._decrypt(ciphertext).then(decrypted => {
+      const { data, message } = decrypted;
+
+      data.channelServerClient = this;
+      data.remoteMetaData = pick(sender, 'city', 'country', 'region', 'ua');
+
+      this.trigger(message, data);
+      this._notifier.trigger(message, data);
+    });
+  }
+
+  /**
+   * Create a WebSocket to `url`
+   *
+   * @param {String} url
+   * @returns {WebSocket}
+   * @private
+   */
+  _createSocket (url) {
+    return new WebSocket(url);
+  }
+
+  /**
+   * Decrypt `ciphertext` using the stored JWK
+   *
+   * @param {String} ciphertext
+   * @returns {Promise} resolves to an object when complete
+   * @private
+   */
+  _decrypt (ciphertext) {
+    return this._getChannelJWK()
+      .then(keysJwk => a256gcm.decrypt(ciphertext, keysJwk))
+      .then(result => JSON.parse(result));
+  }
+
+  /**
+   * Encrypt `envelope` using the stored JWK
+   *
+   * @param {Object} envelope
+   * @returns {Promise} Resolves to a string when complete
+   * @private
+   */
+  _encrypt (envelope) {
+    return this._getChannelJWK()
+      .then(keysJwk => a256gcm.encrypt(JSON.stringify(envelope), keysJwk));
+  }
+
+  /**
+   * Get the encryption/decryption JWK for the channel
+   *
+   * @returns {Promise} resolves to the JWK when complete
+   * @private
+   */
+  _getChannelJWK () {
+    const { channelId, channelJWK, channelKey } = this.toJSON();
+    if (channelJWK) {
+      return Promise.resolve(channelJWK);
+    }
+
+    const channelKeyBuffer = Buffer.from(base64url.decode(channelKey), 'utf8');
+    const channelIdBuffer = Buffer.from(channelId, 'utf8');
+
+    return Promise.all([
+      this._deriveChannelJWK(channelKeyBuffer, channelIdBuffer),
+      this._deriveConfirmationCode(channelKeyBuffer, channelIdBuffer),
+    ]).then(([channelJWK, confirmationCode]) => {
+      this.set({
+        channelJWK,
+        confirmationCode,
+      });
+      return channelJWK;
+    });
+  }
+
+  /**
+   * Use `channelKeyBuffer` and `channelIdBuffer` to derive
+   * the encryption/decryption JWK for the channel
+   *
+   * @param {Buffer} channelKeyBuffer
+   * @param {Buffer} channelIdBuffer
+   * @returns {Promise} resolves to the JWK when complete
+   * @private
+   */
+  _deriveChannelJWK(channelKeyBuffer, channelIdBuffer) {
+    return hkdf(channelKeyBuffer, channelIdBuffer, CHANNEL_KEY_INFO_BUFFER, DEVICE_PAIRING_CHANNEL_KEY_BYTES)
```

Comment:
@rfk, @eoger - is this how you'd expect the channel JWK and confirmation code to be derived?

Reply:
LGTM